File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols (VB)/IConfigurationCellParsingState.vb
'*******************************************************************************************************
'  IConfigurationCellParsingState.vb - Configuration cell parsing state interface
'  Copyright © 2008 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2008
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  04/16/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

''' <summary>This interface represents the protocol independent parsing state of a set of configuration related data settings (typically related to a PMU).</summary>
<CLSCompliant(False)> _
Public Interface IConfigurationCellParsingState

    Inherits IChannelCellParsingState

    Delegate Function CreateNewDefinitionFunctionSignature(Of T As IChannelDefinition)(ByVal parent As IConfigurationCell, ByVal binaryImage As Byte(), ByVal startIndex As Int32) As T

    ReadOnly Property CreateNewPhasorDefintionFunction() As CreateNewDefinitionFunctionSignature(Of IPhasorDefinition)

    ReadOnly Property CreateNewFrequencyDefintionFunction() As CreateNewDefinitionFunctionSignature(Of IFrequencyDefinition)

    ReadOnly Property CreateNewAnalogDefintionFunction() As CreateNewDefinitionFunctionSignature(Of IAnalogDefinition)

    ReadOnly Property CreateNewDigitalDefintionFunction() As CreateNewDefinitionFunctionSignature(Of IDigitalDefinition)

End Interface

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/IConfigurationCellParsingState.cs
//*******************************************************************************************************
//  IConfigurationCellParsingState.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  04/16/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Defines function signature for creating new <see cref="IChannelDefinition"/> objects.
    /// </summary>
    /// <param name="parent">Reference to parent <see cref="IConfigurationCell"/>.</param>
    /// <param name="binaryImage">Binary image to parse <see cref="IChannelDefinition"/> from.</param>
    /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
    /// <returns>New <see cref="IChannelDefinition"/> object.</returns>
    /// <typeparam name="T">Specific <see cref="IChannelDefinition"/> type of object that the <see cref="CreateNewDefinitionFunction{T}"/> creates.</typeparam>
    public delegate T CreateNewDefinitionFunction<T>(IConfigurationCell parent, byte[] binaryImage, int startIndex)
        where T : IChannelDefinition;

    /// <summary>
    /// Represents a protocol independent interface representation of the parsing state of any kind of <see cref="IConfigurationCell"/>.
    /// </summary>
    public interface IConfigurationCellParsingState : IChannelCellParsingState
    {
        /// <summary>
        /// Gets reference to <see cref="CreateNewDefinitionFunction{T}"/> delegate used to create new <see cref="IPhasorDefinition"/> objects.
        /// </summary>
        CreateNewDefinitionFunction<IPhasorDefinition> CreateNewPhasorDefinition { get; }

        /// <summary>
        /// Gets reference to <see cref="CreateNewDefinitionFunction{T}"/> delegate used to create new <see cref="IFrequencyDefinition"/> objects.
        /// </summary>
        CreateNewDefinitionFunction<IFrequencyDefinition> CreateNewFrequencyDefinition { get; }

        /// <summary>
        /// Gets reference to <see cref="CreateNewDefinitionFunction{T}"/> delegate used to create new <see cref="IAnalogDefinition"/> objects.
        /// </summary>
        CreateNewDefinitionFunction<IAnalogDefinition> CreateNewAnalogDefinition { get; }

        /// <summary>
        /// Gets reference to <see cref="CreateNewDefinitionFunction{T}"/> delegate used to create new <see cref="IDigitalDefinition"/> objects.
        /// </summary>
        CreateNewDefinitionFunction<IDigitalDefinition> CreateNewDigitalDefinition { get; }
    }
}